Open Review of Robot Framework® Certified Professional Syllabus (#39)
robotframework/robotframework-RFCP-syllabus: Chapter_4_Advanced_Structuring_and_Execution.md

```

This command will execute tests|tasks that have the `regression` tag but exclude any that also have the `unstable` tag.

#### 4.4.2.4 Using Tag Patterns

Tags can include patterns using wildcards `*` and `?` to match multiple tags:

- `*` matches any number of characters.
- `?` matches any single character.

Examples:
- Include tests|tasks with tags starting with `feature-`:

  ```shell
  robot --include feature-* path/to/tests
  ```

- Exclude tests|tasks with tags ending with `-deprecated`:

  ```shell
  robot --exclude *-deprecated path/to/tests
  ```

### 4.4.3 Reserved Tags

Tags starting with `robot:` are reserved for internal use by Robot Framework and should not be used in user-defined tags.
Using own tags with this prefix may lead to unexpected behavior in test execution and reporting.

- `robot:exclude`: Marks tests|tasks that should be excluded from execution similar to `--exclude`.
- `robot:skip`: Marks tests|tasks that should be skipped during execution similar to `--skip`.

## 4.5 SKIP Test|Task Status

> [!IMPORTANT]
> LO-XX Recall the use-case and purpose of skipping tests|tasks in Robot Framework (K1)
>

> > LO-XX Recall the different ways to skip tests|tasks in Robot Framework (K1)

Reviewer: Make this a K2 because the student must also be able to differentiate them.

In addition to `PASS` and `FAIL`, Robot Framework introduces a `SKIP` status to indicate that a test|task was explicitly skipped **during** execution. The `SKIP` status is useful when certain tests|tasks should not be executed, for example, due to unfulfilled preconditions, unfinished test logic, or unsupported environments. Skipped tests|tasks appear in logs and reports, clearly marked as skipped.

**Reasons to Use SKIP**

- **Temporal Exclusion of Tests|Tasks**: To prevent failing tests|tasks for known issues to run until the issue is resolved.
- **Conditional Execution**: Skip tests|tasks dynamically based on runtime conditions, i.e. if Suite Setup detected an issue.
- **Unsupported Scenarios**: Mark tests|tasks as skipped in environments where they cannot run, but shall be in logs.

### 4.5.1 Skipping By Tags Selection (CLI)

> [!IMPORTANT]

> > LO-XX Recall the differences between skip and exclude (K1)

Reviewer: Also here: to weigh them up against each other is a K2 skill.

A **Setup** is a single keyword with potential argument values that is called before all other keywords; or before tests|tasks in Suite Setup.

Examples of typical use cases for Setups are:
- Establishing connections to databases or services.
- Initializing test data or configurations.
- Setting the system under test to a known state.
- Logging into applications or systems.
- Navigating to the feature under test.

### 4.1.1 Suite Setup

> [!IMPORTANT]
> LO-XX Recall key characteristics, benefits, and syntax of Suite Setup (K1)
>
> LO-XX Understand when Suite Setup is executed and used (K2)

A **Suite Setup** is executed before any tests|tasks or child suites within the suite are run.
It is used to prepare the environment or perform actions that need to occur before the entire suite runs.
Since it is only executed once before all tests|tasks or child suites, it can save time, rather than executing the action for each test|task individually.

**Key characteristics of Suite Setup:**
- Executed before any tests|tasks and child suites in the suite.
- If the Suite Setup fails, all tests|tasks in the suite and its child suites are marked as failed, and they are not executed.

> - Ideal for checking **preconditions** that must be met before running the tests|tasks.

Reviewer 2: To me this feels more like a use cases instead of a characteristic

Author: True

- Logged in the execution log as a separate section, indicating the setup status.

**Typical use cases:**
- Ensuring that the environment is ready for execution.
- Starting services or applications required for the suite.

> - Preparing a system under test to meet the suite's requirements.

Reviewer 2: I would leave out "under test" as this focusses on tests, but it's the same for tasks

```suggestion
- Preparing a system to meet the suite's requirements.
```

### 4.1.2 Test|Task Setup

> [!IMPORTANT]
> LO-XX Recall key characteristics, benefits, and syntax of Test Setup (K1)
>
> LO-XX Understand when Test|Task Setup is executed and used (K2)

A **Test|Task Setup** is executed before a single test|task runs.
It is used to prepare the specific conditions required for that test|task.

You can define a default Test|Task Setup in the `*** Settings ***` section of the suite using the `Test Setup`|`Task Setup` setting.
This setup will be applied to all tests|tasks within the suite unless overridden.

Individual tests|tasks can override the default setup by specifying their own `[Setup]` setting within the test|task.
To disable the setup for a specific test|task, you can set `[Setup]    NONE`, which means that no setup will be executed for that test|task.

**Key characteristics of Test|Task Setup:**
- Executed before the test|task starts.
- If the Test|Task Setup fails, the test|task is marked as failed, and its body, including its main keywords, is not executed.
- Can be set globally for all tests|tasks in a suite and overridden locally.
- Logged in the execution log as a separate section, indicating the setup status.

**Typical use cases:**

> - Setting up test data unique to the test|task.

Reviewer 2: leave out test in the first part of the sentence
```suggestion
- Setting up data unique to the test|task.
```

> - Distinguishing keywords of a test.

Author: i clarified it a bit:
- Distinguishing phases of a test|task in *setup* (aka *preparation* or *precondition checking*), steps, and teardown (aka *clean up* or *postconditions*).
@frankvanderkuur

> - Executing preparation steps to navigate to the feature under test.

Reviewer 2: Don't make this test specific, because it could be the same for tasks:

```suggestion
- Execute preparation steps to navigatie to the test feature/task
```

Author: No i disagree with the grammar.
And here i think this is kind of specific to testing.
@frankvanderkuur

Reviewer 2: Grammar-wise I'm fine with whatever solution. Although I don't have heaps of experience with RPA, I can see situations where you could benefit from using a task setup. But, I wouldn't say this needs a whole lot of debating. In or out, either is fine with me.

Author: changed

    [Setup]    Open Data Connection
    Process the Data
```

## 4.2 Teardowns (Suite, Test|Task, Keyword)

> [!IMPORTANT]
> LO-XX Understand the different levels where and how a Teardowns can be defined and when they are executed (K2)
>
> LO-XX Recall the typical use cases for using a Teardowns (K1)

In automation, tests|tasks are typically executed in a linear sequence.
This linear execution can lead to issues when a preceding test|task fails, potentially affecting subsequent tests|tasks due to an unclean state of the system under test or the automated environment.
To prevent such issues, Robot Framework provides the **Teardown** functionality, which can be defined at the suite, test|task, or keyword level.

As mentioned before, a failure resulting in a keyword with the status `FAIL` will cause Robot Framework not to execute all subsequent keywords of the current test|task.
These not-executed keywords will receive the status `NOT RUN`.

A **Teardown** is a single keyword call with potential argument values that is executed after the child suites, test|tasks, and keywords have completed execution, regardless of the outcome, even if previously executed elements have failed.

> It ensures that necessary cleanup actions are performed, maintaining the integrity of the test environment for subsequent executions.

Reviewer 2: leave out test, because RPA would be done in a production environment
```suggestion
It ensures that necessary cleanup actions are performed, maintaining the integrity of the environment for subsequent executions.
```

    Perform Admin Actions   007

No Setup Test
    [Setup]    NONE    # Override and disable the setup by case-sensitive NONE
    Perform Actions Without Login   000
```

### 4.1.3 Keyword Setup

> [!IMPORTANT]
> LO-XX Recall key characteristics and syntax of Keyword Setup (K1)

A **Keyword Setup** is executed before the body of a user keyword is executed.
It allows for preparation steps specific to that keyword or ensures that the keyword's requirements are met before execution.

**Key characteristics of Keyword Setup:**
- Keyword Setup is a single keyword call with potential argument values.
- Executed before the keyword's body.
- If the Keyword Setup fails, the keyword's body is not executed.
- Logged in the execution log as a separate section, indicating the setup status.

**Typical use cases:**
- Opening connections or files needed by the keyword.
- Initializing variables or data structures.
- Ensuring preconditions specific to the keyword are met.

Example of defining a Keyword Setup:

```robotframework
*** Keywords ***
Process Data
    [Setup]    Open Data Connection
    Process the Data
```

## 4.2 Teardowns (Suite, Test|Task, Keyword)

> [!IMPORTANT]

> > LO-XX Understand the different levels where and how a Teardowns can be defined and when they are executed (K2)

Reviewer: ```suggestion
> LO-XX Understand the different levels where and how Teardowns can be defined and when they are executed (K2)
```

>

> > LO-XX Recall the typical use cases for using a Teardowns (K1)

Reviewer: ```suggestion
> LO-XX Recall the typical use cases for using Teardowns (K1)
```
